micro_sam/sam_3d_wrapper.py

```diff
@@ -0,0 +1,151 @@
+    ):
+        super().__init__(sam_model)
+        self.model_type = model_type
+        self.d_size = d_size
+
+        # predictor.model = Sam3DWrapper(predictor.model, self.d_size)
+        # self.predictor = predictor
+
+
+class Sam3DWrapper(nn.Module):
+    def __init__(
+        self,
+        sam_model: Sam,
+        d_size,
+    ):
+        super().__init__()
+        self.sam_model = sam_model
+        self.d_size = d_size
+        sam_model.image_encoder = ImageEncoderViT3DWrapper(image_encoder=self.sam_model.image_encoder, d_size=self.d_size)
+        self.sam_model = sam_model
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.sam_model(x, d_size=self.d_size)
+
+
+class ImageEncoderViT3DWrapper(nn.Module):
+    def __init__(
+        self,
+        image_encoder: nn.Module,
+        num_heads: int = 12,
+        embed_dim: int = 768,
+        **kwargs
+    ):
+        super().__init__()
+        self.image_encoder = image_encoder
+        self.img_size = self.image_encoder.img_size
+
+        # freeze sam blocks
+        for k, v in self.image_encoder.named_parameters():
+            if '.adapter_' not in k:
+                v.requires_grad = False
+
+        # replace default blocks with 3d adapter blocks
+        for i, blk in enumerate(self.image_encoder.blocks):
+            self.image_encoder.blocks[i] = NDBlock(Block=blk, num_heads=num_heads, dim=embed_dim)
+
+    def forward(self, x: torch.Tensor, d_size) -> torch.Tensor:
+        x = self.image_encoder.patch_embed(x)
+        if self.image_encoder.pos_embed is not None:
+            x = x + self.image_encoder.pos_embed
+
+        for blk in self.image_encoder.blocks:
+            x = blk(x, d_size)
+
+        x = self.image_encoder.neck(x.permute(0, 3, 1, 2))
+
+        return x
+
+
+class NDBlock(nn.Module):
+    def __init__(
+            self,
+            dim: int,
+            num_heads: int,
+            Block: nn.Module,
+            norm_layer: Type[nn.Module] = nn.LayerNorm,
+            **kwargs
+    ):
+        super().__init__()
+        self.Block = Block
+        
+        self.adapter_channels = 384
+        self.adapter_linear_down = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act = nn.GELU()
+        self.adapter_norm = norm_layer(dim)
+
+        self.adapter_linear_down_2 = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up_2 = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv_2 = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act_2 = nn.GELU()
+        self.adapter_norm_2 = norm_layer(dim)
+
+    def forward(self, x: torch.Tensor, d_size) -> torch.Tensor:
```

Comment:
It seems like this was never called, otherwise you would have run into problems due to `self.attn` etc. not defined.